### PR TITLE
Feature/profile page builder

### PR DESCRIPTION
**Pull request by bryanloh-jw**
# Profile Page Builder Feature

## Context

This feature is for users to customize their profile page that visitors to their page will see. Basic information will be required, and additional information such as socials, links, files, etc will be available for users to fully showcase themselves through their page. The application will require an easy to use interface that will enable users to easily add what they require, and provide elements of customization for them.

### Banner & Profile Image

Both the banner and profile image can take in uploads for picture files to be displayed. The images received from the user will be temporarily stored on the browser and referenced to be displayed. Currently the users have no way to adjust the display of the images, but that can be added in the future.

For the banner, a color picker is also implemented using the library 'react-color'. Selecting a color will overwrite the display picture and vice-versa.

### Core information

The core information for user profile is a few inputs that will be required for the basic profile. They will always be displayed on the top of the profile edit page. The information is stored in useState in the EditProfile component.

### Widgets

One creative feature of the edit profile page is to have widgets that hold different information for the user. They can be added and customized to the user's information. Currently, socials and link widgets have been implemented.

The widget components are made generic by using the WidgetItem component and passing a type along with a value. The type corresponds to different widgets and will get the component tied to it. Values or states being passed to the components are stringified values that contains user input. Zod library is used to ensure that the values being passed to each component matches the type required to be used.

The state of all the widgets are within the EditProfile component. Application wide state management was initially considered to hold the state of the profile builder, however it would be overkill as the state can be easily managed with minimal prop drilling and the passing of functions as props to edit state within the individual widgets. The state holds an array of widget information in the order that will be displayed. Each object holds the widget type, the stringified values, and unique id that helps with rendering and the sorting library.

A sorting / drag and drop feature was implemented using 'react-easy-sort' library. Initially 'sortablejs' was used but the dragging feature was not easy to implement, and the library was not mature. 'react-easy-sort' library fulfils the requirement of allowing widgets to be reordered by dragging and dropping.

## Future works

Data persistence for profile and banner images.

Form submission for all the user input.

Additional widgets for users to choose from.

### CODE REVIEW

File: the-name-card/src/components/EditProfile/EditProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,309 @@
+import { ChangeEvent, useRef, useState } from 'react';
```
Comment (raynerljm) [inline the-name-card/src/components/EditProfile/EditProfile.tsx]: might want to put in this component in another directory (maybe called `pages`?) to distinguish components from pages

File: the-name-card/src/components/EditProfile/EditProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+	WidgetInitialValue,
+	WidgetItem,
+	WidgetProp,
+	WidgetType,
+} from '..';
+import './EditProfile.scss';
+
+export const EditProfile = () => {
+	// banner stuff for future refactor
+	const [bannerColor, setBannerColor] = useState<string>('#10A5F5');
```
Comment (raynerljm) [inline the-name-card/src/components/EditProfile/EditProfile.tsx]:
```suggestion
	const [bannerColor, setBannerColor] = useState('#10A5F5');
```
unnecessary type declartaion

File: the-name-card/src/components/EditProfile/EditProfile.tsx (hunk 1 above)
Comment (raynerljm) [reply the-name-card/src/components/EditProfile/EditProfile.tsx]: nit: also this file is a bit long (~300 lines) if you want to follow the good practice of keeping components leaner (100-200 lines) u might want to refactor out some parts

File: the-name-card/src/components/EditProfile/EditProfile.tsx [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+									type="text"
+								/>
+							</div>
+						</div>
+					</div>
+					<SortableList
+						onSortEnd={onSortEnd}
+						draggedItemClassName="highlight"
+					>
+						{widgetProperties.map((widget, index) => {
```
Comment (raynerljm) [inline the-name-card/src/components/EditProfile/EditProfile.tsx]: good shit

File: the-name-card/src/components/EditProfile/EditProfile.tsx [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+	const openProfileFileInput = () => {
+		profilePictureInputRef.current?.click();
+	};
+
+	const onSortEnd = (oldIndex: number, newIndex: number) => {
+		setWidgetProperties((array) => arrayMove(array, oldIndex, newIndex));
+	};
+
+	const setWidgetPropertiesWrapper = (
```
Comment (raynerljm) [inline the-name-card/src/components/EditProfile/EditProfile.tsx]: personally, since this function is returning a function, i would name it smth else to make that clear
e.g. `generateWidgetPropertiesSetter` or `generateSetWidgetProperties`)

File: the-name-card/src/components/WidgetItem/Items/Link.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import { WidgetProp } from '.';
+import './Link.scss';
+
+const LinkValues = z.object({ description: z.string(), link: z.string() });
+
+export type LinkValues = z.infer<typeof LinkValues>;
+
+export const Link: React.FC<WidgetProp> = ({ value, updateValue }) => {
+	const rawValue = JSON.parse(value);
+	const linkValues = LinkValues.parse(rawValue);
```
Comment (raynerljm) [inline the-name-card/src/components/WidgetItem/Items/Link.tsx:12]: this might result in the page crashing because there's no error handler if `JSON.parse` or `LinkValues.parse` fails

File: the-name-card/src/components/WidgetItem/Items/Link.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+	const editDescription = (
+		type: 'description' | 'link',
```
Comment (raynerljm) [inline the-name-card/src/components/WidgetItem/Items/Link.tsx:15]: nit: possible to abstract out these types somewhere

File: the-name-card/src/components/WidgetItem/Items/Socials.scss [hunk 7]
```diff
@@ -0,0 +1,96 @@
+@import './../../../styles/variables';
```
Comment (raynerljm) [inline the-name-card/src/components/WidgetItem/Items/Socials.scss]: might want to look into path aliases (for both ur `.ts` and `.scss` imports)

File: the-name-card/src/components/WidgetItem/Items/Socials.tsx [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	},
+	[SocialMediaType.TikTok]: {
+		icon: FaTiktok,
+		url: 'www.tiktok.com/@',
+	},
+};
+
+export const Socials: React.FC<WidgetProp> = ({ value, updateValue }) => {
+	const rawValue = JSON.parse(value);
+	let socialsValue = SocialArray.parse(rawValue);
```
Comment (raynerljm) [inline the-name-card/src/components/WidgetItem/Items/Socials.tsx:73]: i'm not 100% sure but idt declaring a variable like that and modifying is standard practice. might have to declare a `useState` instead (but might be better to google)

File: the-name-card/src/components/WidgetItem/Items/Socials.tsx [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+	const [existingSocials, setExistingSocials] = useState<
+		Set<SocialMediaType>
+	>(new Set());
+	const [showOptions, setShowOptions] = useState(false);
+
+	useEffect(() => {
```
Comment (raynerljm) [inline the-name-card/src/components/WidgetItem/Items/Socials.tsx:79]: ur `useEffect` needs a dependency array

File: the-name-card/src/components/WidgetItem/Items/Socials.tsx [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+		url: 'www.youtube.com/@',
+	},
+	[SocialMediaType.TikTok]: {
+		icon: FaTiktok,
+		url: 'www.tiktok.com/@',
+	},
+};
+
+export const Socials: React.FC<WidgetProp> = ({ value, updateValue }) => {
+	const rawValue = JSON.parse(value);
```
Comment (raynerljm) [inline the-name-card/src/components/WidgetItem/Items/Socials.tsx:72]: same thing here about no error handling for incorrect parsing

**Review by raynerljm — COMMENTED** [review]
left some comments

File: the-name-card/src/components/EditProfile/EditProfile.tsx (hunk 1 above)
Comment (bryanloh-jw) [reply the-name-card/src/components/EditProfile/EditProfile.tsx]: on both points, will be doing a major refactoring for soon.

for the pages directory will be pulling the profile page & landing page out and placing their components in the aptly named component directory

as for the EditProfile component, the idea was to get the whole component working to solely focus on functionality before splitting to different components. As the component is quite logic heavy, I thought that it would be easier to work in a single file at first.

tldr: yes will be refactoring for maintainability and clarity

File: the-name-card/src/components/EditProfile/EditProfile.tsx (hunk 4 above)
Comment (bryanloh-jw) [reply the-name-card/src/components/EditProfile/EditProfile.tsx]: sounds good, I rarely work with higher-order functions so didn't know about the naming convention

File: the-name-card/src/components/WidgetItem/Items/Link.tsx (hunk 5 above)
Comment (bryanloh-jw) [reply the-name-card/src/components/WidgetItem/Items/Link.tsx:12]: yeap, will be working on the error handling for this

File: the-name-card/src/components/WidgetItem/Items/Link.tsx (hunk 6 above)
Comment (bryanloh-jw) [reply the-name-card/src/components/WidgetItem/Items/Link.tsx:15]: will do
